### PR TITLE
[18.06]numpy for Python

### PR DESCRIPTION
**Pull request by jpc**
Maintainer: @commodo
Compile & run tested: OpenWRT 18.06 target-aarch64_cortex-a53_musl NanoPi Neo Core 2

Description:
These are some pretty big hacks required to cross compile numpy. This required both changes to python distutils and hacking platform detection in numpy compilation scripts.

Issues:
1. It compiles and simple use cases work but I have not tested it extensively yet (I did not run the test suite on the device).
2. python2 will probably need a similar distutils fix (I did not need python2 support so I did touch this)
3. I am not sure if this should be merged as is but maybe somebody finds it useful (I would) while we figure out a way to make it less hacky.

### CODE REVIEW

**Comment by jpc** [comment]
I guess this will not be merged into 18.06 because of the feature freeze but I would still love to discuss how to attack this problem.

**Comment by commodo** [comment]
Hey,

I’m on vacation until September.
I’m not sure if I will  be able to look closely earlier.

The Python3 patch looks fine from my side.
It might make sense to upstream this to Python.
If they accept this, then it’s simple for us.

Does numpy have a tar.gz package?
I usually prefer those, since .zip require special handling in some cases.

Regarding the numpy patch, that looks ok overall.
I may have just one comment. I’ll try to add it.
But again, if you have something that could be sent upstream to numpy, it would also be great.

**Comment by commodo** [comment]
Thanks for the package :)
(I forgot to say this earlier)

It’s been a while since used numpy, and it’s one of the packages that (I feel) makes Python really neat.

File: lang/python/python-package.mk [hunk 1]
```diff
@@ -94,6 +94,9 @@ define Build/Compile/HostPyRunTarget
 		CXX="$(TARGET_CXX)" \
 		LD="$(TARGET_CC)" \
 		LDSHARED="$(TARGET_CC) -shared" \
+		F90="$(TARGET_CROSS)gfortran" \
+		AR="$(TARGET_AR)" \
+		RANLIB="$(TARGET_RANLIB)" \
```
Comment (jefferyto) [inline lang/python/python-package.mk]: Would prefer to set these using `PYTHON[3]_PKG_SETUP_VARS`, unless there are other packages that also need this.

File: lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+         if 'ARFLAGS' in os.environ:
+             archiver = ar + ' ' + os.environ['ARFLAGS']
+         else:
+@@ -215,7 +217,8 @@ def customize_compiler(compiler):
+             compiler_cxx=cxx,
+             linker_so=ldshared,
+             linker_exe=cc,
+-            archiver=archiver)
++            archiver=archiver,
++            ranlib=ranlib)
```
Comment (jefferyto) [inline lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch]: Would prefer to find another way to solve this, again unless there is more than one package that requires patching Python directly.

From [this](https://github.com/numpy/numpy/blob/v1.16.3/numpy/distutils/fcompiler/__init__.py#L121) I would assume it looks for a `RANLIB` environment variable?

Comment (jpc) [reply lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch]: I don't remember all the specifics now but I spent several hours io:stderr:write'ing through the code to hunt down where the ranlib is passed through to the actual executed commands. The code and data flow jumped through files and classes and was very hard to follow.

I've sent this patch to Python, we can wait a bit and see what they think about this (https://bugs.python.org/user32173)

File: lang/python/python-package.mk (hunk 1 above)
Comment (jpc) [reply lang/python/python-package.mk]: Are you sure? Fortran seems pretty standard to me and is probably used in more scientific libraries. It should not break anything in cases where the Fortran compiler is not used.

As far as ar and ranlib are concerned it may be unnecessary on Linux (I don't think ARCH makes a difference for these binutils tools). On macOS ar is platform agnostic but ranlib is not (and the error one gets is not very enlightening). In any case it is safer to use proper cross-tools and adding it for all packages may save someone some headaches in the future.

Comment (jefferyto) [reply lang/python/python-package.mk]: I've grepped packages in the base and packages (this one) feeds, not a single package uses Fortran.

I know you were developing with 18.06 which doesn't have [PYTHON[3]_PKG_SETUP_VARS](https://github.com/openwrt/packages/tree/master/lang/python#building), but if you were to submit a PR for master, I would suggest you use it.

Comment (commodo) [reply lang/python/python-package.mk]: I think `AR` & `RANLIB` would be fine to be set here.
They look like they could be beneficial for other packages with macOS; they could be made part of the "python cross-build infrastructure".
Since I don't have a Mac machine, cross-compiling on a Mac for me is a black-box.

Whereas fortran, I personally don't mind; it looks more like a relic [I agree].
No strong opinion on it.

In general, I am fine with taking some risks every now-n-then.
This looks like one of those moments where things look weird enough that the first instinct is to resist them.

Well, this is my vote about this :)

File: lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch (hunk 2 above)
Comment (commodo) [reply lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch]: the Python thread seems to be a bit slow;
maybe it's vacation time for more people;

let's ping it

Comment (commodo) [reply lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch]: oh; regarding [the PR](https://github.com/python/cpython/pull/15387), you need to sign a [Python CLA](https://www.python.org/psf/contrib/contrib-form/)

that's some legal stuff that they need, because ¯\_(ツ)_/¯

Comment (jpc) [reply lang/python/python3/patches/020-pass-cross-ranlib-through-distutils.patch]: Yeah, I noticed and it is already tagged as having the CLA. The numpy maintainer offered to help with the process if nobody steps up so maybe I should take him up on that offer. :)

File: lang/python/python-package.mk (hunk 1 above)
Comment (jpc) [reply lang/python/python-package.mk]: Yeah, I forgot that Fortran is a numpy-specific addition to distutils so it totally makes sense to move it to the numpy package using PYTHON_PKG_SETUP_VARS.
I think `AR` and `RANLIB` make more sense for packages other than numpy so I think it is a good idea to keep it.

Comment (jefferyto) [reply lang/python/python-package.mk]: I'll defer to @commodo on adding `AR` and `RANLIB`.

But to beat a dead horse one last time: I believe in the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)). If there are other existing packages need similar workarounds for `AR` and/or `RANLIB`, then I'd be happy to refactor this code into python-package.mk. But if there isn't, then we are adding code that has no immediate benefit but increases the potential for error (across all Python packages).

Comment (commodo) [reply lang/python/python-package.mk]: i'm more inclined to stick with AR & RANLIB as proposed by @jpc

to some degree, I feel that some of these `TARGET_` options should be similar or identical to `TARGET_CONFIGURE_OPTS` as here https://github.com/openwrt/openwrt/blob/master/rules.mk#L302

macOS is a bit special sometimes during cross-building (especially when building OpenWrt) in the fact that it reveals weird stuff; and not many people try to build OpenWrt on macOS;
and i've also wasted time to fix mac-build stuff (when I had one around at work)

sometimes, I look at buildroot for inspiration (because OpenWrt is an older fork of buildroot), but that still doesn't offer any help in this AR + RANLIB discussion;

personally, i am willing to take a chance with this change, so if there's any breakage with it, we can probably revert;

one thing does come to mind here though: maybe i should buy a Mac as personal computer  ¯\_(ツ)_/¯

**Comment by bhack** [comment]
Has this package PR been abandoned?

**Comment by commodo** [comment]
Funny story.
I was thinking about packaging this a few days ago for something I need.
I forgot about this PR.

Will try to pick it up in the next few days.

**Comment by bhack** [comment]
@commodo If you can, ping here when you will have something so we are notified. Thanks.

**Comment by bhack** [comment]
Have openwrt any accellerated Blas library package (Atlas or Openblas) so that the numpy package can depend on?

**Comment by bhack** [comment]
I've a very Draft openblas package PR at https://github.com/openwrt/packages/pull/11894
